Twitter is not filtered by content filter

pvollan opened the pull request:
#### 78e6b51bca3333ed35eb4c3448440f52c565be38
<pre>
Twitter is not filtered by content filter
<a href="https://bugs.webkit.org/show_bug.cgi?id=252496">https://bugs.webkit.org/show_bug.cgi?id=252496</a>
rdar://103032824

Reviewed by Sihui Liu.

Twitter is not filtered by content filter since service worker fetch is missing filtering code.
Make sure content filtering is started on service worker fetch, and let the content filter
intercept the fetch if it should be blocked. The didReceiveData method in ServiceWorkerFetchTask
has been split into two methods, since it was being called both when receiving data over IPC and
from the network. The unsafeBuffer method in IPC::SharedBufferReference is only allowed to be
called on the receiver side. The call to startContentFiltering was moved to a location where it
will also be called on service worker fetch, which required a little refactoring.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/service-workers/basic-fetch-with-contentfilter.https-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/basic-fetch-with-contentfilter.https.html: Added.
* LayoutTests/http/wpt/service-workers/basic-fetch-with-contentfilter.js: Added.
(doTest):
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::start):
(WebKit::NetworkResourceLoader::startRequest):
(WebKit::NetworkResourceLoader::startWithServiceWorker):
(WebKit::NetworkResourceLoader::requestIsForServiceWorkerOnly):
(WebKit::NetworkResourceLoader::serviceWorkerDidNotHandle):
(WebKit::NetworkResourceLoader::continueAfterServiceWorkerReceivedData):
(WebKit::NetworkResourceLoader::continueAfterServiceWorkerReceivedResponse):
(WebKit::NetworkResourceLoader::serviceWorkerDidFinish):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::processResponse):
(WebKit::ServiceWorkerFetchTask::didReceiveData):
(WebKit::ServiceWorkerFetchTask::didReceiveDataFromPreloader):
(WebKit::ServiceWorkerFetchTask::didFinish):
(WebKit::ServiceWorkerFetchTask::loadBodyFromPreloader):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.h:

Canonical link: <a href="https://commits.webkit.org/260891@main">https://commits.webkit.org/260891@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ecd6faab96dbce6a8665ef4b70c8c3926383e89c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109828 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18941 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42557 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1290 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20408 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10137 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102098 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115576 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43415 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30079 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85215 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11660 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/12304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51030 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7551 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14066 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
```diff
@@ -208,6 +213,10 @@ void ServiceWorkerFetchTask::didReceiveResponse(WebCore::ResourceResponse&& resp
         cancelPreloadIfNecessary();
 
+#if ENABLE(CONTENT_FILTERING_IN_NETWORKING_PROCESS)
+    m_loader.continueAfterResponseReceived(response);
```
Shouldn't we check return value here?

pvollan replied:
> Shouldn't we check return value here?

Yes, that is a good catch. Fixed in latest patch.

Thanks for reviewing!

szewai commented on Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
```diff
@@ -254,8 +263,29 @@ void ServiceWorkerFetchTask::didReceiveData(const IPC::SharedBufferReference& da
+    sendToClient(Messages::WebResourceLoader::DidReceiveData { IPC::SharedBufferReference(buffer), encodedDataLength });
+}
+
+void ServiceWorkerFetchTask::didReceiveDataFromPreloader(const WebCore::FragmentedSharedBuffer& data, uint64_t encodedDataLength)
```
Why do we need to break down `ServiceWorkerFetchTask::didReceiveData` into two functions? Is it because the buffer needs to be contiguous? `SharedBufferReference::unsafeBuffer()` seems can do that.

pvollan replied:
> Why do we need to break down `ServiceWorkerFetchTask::didReceiveData` into two functions? Is it because the buffer needs to be contiguous? `SharedBufferReference::unsafeBuffer()` seems can do that.

The didReceiveData method in ServiceWorkerFetchTask has been split into two methods, since it was being called both when receiving data over IPC and from the network. The unsafeBuffer method in IPC::SharedBufferReference is only allowed to be called on the receiver side.

szewai reviewed (commented):
I think it's better to include a test; or explanation in the commit message that why a test cannot be added (e.g. feature is not enabled, test infra does not support, etc) and how this change is verified.

szewai reviewed (approved):
r=me

szewai commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -1899,26 +1898,44 @@ void NetworkResourceLoader::setWorkerStart(MonotonicTime value)
+    if (m_parameters.serviceWorkersMode != ServiceWorkersMode::Only)
+        return false;
+
+    LOADER_RELEASE_LOG_ERROR("serviceWorkerDidNotHandle: Aborting load because the service worker did not handle the load and serviceWorkerMode only allows service workers");
```
It looks like this log should mention `requestIsForServiceWorkerOnly` instead of `serviceWorkerDidNotHandle`.

szewai commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -1899,26 +1898,44 @@ void NetworkResourceLoader::setWorkerStart(MonotonicTime value)
+    startRequest(newRequest);
+}
+
+bool NetworkResourceLoader::requestIsForServiceWorkerOnly()
```
It may be named as `checkServiceWorkersMode` or `abortIfServiceWorkersOnly`, since `requestIsForServiceWorkerOnly` sounds like it only returns the value of `m_parameters.serviceWorkersMode == ServiceWorkersMode::Only`.

pvollan replied:
> It may be named as `checkServiceWorkersMode` or `abortIfServiceWorkersOnly`, since `requestIsForServiceWorkerOnly` sounds like it only returns the value of `m_parameters.serviceWorkersMode == ServiceWorkersMode::Only`.

Fixed!

szewai commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -191,16 +191,15 @@ bool NetworkResourceLoader::isSynchronous() const
+void NetworkResourceLoader::startRequest(const ResourceRequest& newRequest)
 {
     ASSERT(RunLoop::isMain());
     LOADER_RELEASE_LOG("start: hasNetworkLoadChecker=%d", !!m_networkLoadChecker);
```
The log may need to be updated to reflect the new function name.

szewai commented on Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
```diff
@@ -1899,26 +1898,44 @@ void NetworkResourceLoader::setWorkerStart(MonotonicTime value)
+    if (requestIsForServiceWorkerOnly())
+        return;
+
+    startRequest(newRequest);
```
Can we just do `serviceWorkerDidNotHandle(nullptr);` here? It seems `start()` in `serviceWorkerDidNotHandle` will filter the content anyways?

pvollan replied:
Thanks for reviewing! I will address these issues in the next patch.

pvollan replied:
> Can we just do `serviceWorkerDidNotHandle(nullptr);` here? It seems `start()` in `serviceWorkerDidNotHandle` will filter the content anyways?

That is a good point. A feature of the content filter is that it can modify the URL being loaded. If we call `serviceWorkerDidNotHandle(nullptr);` here, we will load the original URL, regardless of whether the content filter modified the URL.

Thanks for reviewing!

szewai reviewed (approved):
r=me with minor comments

pvollan committed:
Twitter is not filtered by content filter
https://bugs.webkit.org/show_bug.cgi?id=252496
rdar://103032824

Reviewed by Sihui Liu.

Twitter is not filtered by content filter since service worker fetch is missing filtering code.
Make sure content filtering is started on service worker fetch, and let the content filter
intercept the fetch if it should be blocked. The didReceiveData method in ServiceWorkerFetchTask
has been split into two methods, since it was being called both when receiving data over IPC and
from the network. The unsafeBuffer method in IPC::SharedBufferReference is only allowed to be
called on the receiver side. The call to startContentFiltering was moved to a location where it
will also be called on service worker fetch, which required a little refactoring.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/service-workers/basic-fetch-with-contentfilter.https-expected.txt: Added.
* LayoutTests/http/wpt/service-workers/basic-fetch-with-contentfilter.https.html: Added.
* LayoutTests/http/wpt/service-workers/basic-fetch-with-contentfilter.js: Added.
(doTest):
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebKit/NetworkProcess/NetworkResourceLoader.cpp:
(WebKit::NetworkResourceLoader::start):
(WebKit::NetworkResourceLoader::startRequest):
(WebKit::NetworkResourceLoader::startWithServiceWorker):
(WebKit::NetworkResourceLoader::requestIsForServiceWorkerOnly):
(WebKit::NetworkResourceLoader::serviceWorkerDidNotHandle):
(WebKit::NetworkResourceLoader::continueAfterServiceWorkerReceivedData):
(WebKit::NetworkResourceLoader::continueAfterServiceWorkerReceivedResponse):
(WebKit::NetworkResourceLoader::serviceWorkerDidFinish):
* Source/WebKit/NetworkProcess/NetworkResourceLoader.h:
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::processResponse):
(WebKit::ServiceWorkerFetchTask::didReceiveData):
(WebKit::ServiceWorkerFetchTask::didReceiveDataFromPreloader):
(WebKit::ServiceWorkerFetchTask::didFinish):
(WebKit::ServiceWorkerFetchTask::loadBodyFromPreloader):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.h:

Canonical link: https://commits.webkit.org/260891@main
